tests/filter-video-average.cpp

```diff
@@ -0,0 +1,190 @@
+    clock_shift_ms(&clock, time_limit_ms);
+    OK(acquire_start(runtime));
+    {
+        uint64_t nframes = 0;
+        const uint64_t expected_nframes = props.video[0].max_frame_count / props.video[0].frame_average_count;
+        LOG("Expecting %d frames", expected_nframes);
+
+        // Each pixel is drawn from a uniform distribution in [0, 255].
+        // Without averaging we would expect the intra-frame pixel value
+        // variance to be 255^2 / 12. By averaging over every two frames,
+        // this shrinks to 255^2 / 24.
+        const double expected_pixel_variance = 2709.375;  // 255 * 255 / 24;
+        const size_t num_pixels = props.video[0].camera.settings.shape.x * props.video[0].camera.settings.shape.y;
+        const double normalization_factor = 1.0 / (num_pixels * expected_nframes);
+        double actual_pixel_mean = 0;
+        double actual_pixel_sum_of_squares = 0;
+
+        while (nframes < expected_nframes) {
+            struct clock throttle
+            {};
+            clock_init(&throttle);
+            EXPECT(clock_cmp_now(&clock) < 0,
+                   "Timeout at %f ms",
+                   clock_toc_ms(&clock) + time_limit_ms);
+            VideoFrame *beg, *end, *cur;
+            OK(acquire_map_read(runtime, 0, &beg, &end));
+            for (cur = beg; cur < end; cur = next(cur)) {
+                LOG("stream %d counting frame w id %d", 0, cur->frame_id);
+                CHECK(cur->shape.dims.width ==
+                      props.video[0].camera.settings.shape.x);
+                CHECK(cur->shape.dims.height ==
+                      props.video[0].camera.settings.shape.y);
+                float * data = (float *)&(cur->data[0]);
+                for (size_t i=0; i < num_pixels; ++i) {
+                    const double value = (double)data[i];
+                    actual_pixel_mean += normalization_factor * value;
+                    actual_pixel_sum_of_squares += normalization_factor * value * value;
+                }
+                ++nframes;
+            }
+            {
+                uint32_t n = (uint32_t)consumed_bytes(beg, end);
+                OK(acquire_unmap_read(runtime, 0, n));
+                if (n)
+                    LOG("stream %d consumed bytes %d", 0, n);
+            }
+            clock_sleep_ms(&throttle, 100.0f);
+
+            LOG("stream %d nframes %d. remaining time %f s",
+                0,
+                nframes,
+                -1e-3 * clock_toc_ms(&clock));
+        }
+
+        CHECK(nframes == expected_nframes);
+        // Our tolerance is very loose since the pixel values are very
+        // high and we're only averaging over every two frames.
+        double actual_pixel_variance = actual_pixel_sum_of_squares - (actual_pixel_mean * actual_pixel_mean);
+        LOGE("pixel variance: actual = %g, expected = %g", actual_pixel_variance, expected_pixel_variance);
+        assert_within_abs(actual_pixel_variance, expected_pixel_variance, 100);
```

Comment:
This was because the wrong expected variance was based on that of a continuous uniform distribution, rather than the discrete uniform being used here.